# Segmenting and Clustering Neighbourhoods in Toronto

In this assignment we will explore the neighbourhoods of Toronto using data from Wikipedia to identify the different neighbourhoods, determining and mapping their locations and exploring and clustering them.

## Part 1

For the first part of the assignment we scrape the Wikipedia page for a list of Toronto postal codes with the associated boroughs and neighbourhoods. We will clean the data, removing any missing data points and store the data in a dataframe.

**Import relevant libraries**

In [1]:
import pandas as pd
import bs4 as bs
import urllib.request
import numpy as np

**Scrape table from website using Beautiful Soup**

In [2]:
source = urllib.request.urlopen('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').read()
soup = bs.BeautifulSoup(source,'lxml')

table = soup.find_all('table')


**Convert table to dataframe**

In [3]:
df = pd.read_html(str(table))[0]
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

**Remove all rows with no assigned borough**

In [4]:
df.replace("Not assigned", np.nan, inplace = True)
df.dropna(subset=["Borough"], axis=0, inplace=True)
df.reset_index(drop=True, inplace=True)
df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

**Print number of rows in the cleaned dataframe**

In [5]:
print("Number of rows:",df.shape[0])

Number of rows: 103


## Part 2 

Secondly we add location data (longitude and latitude) for each of the postal codes to the dataframe.

**Add location data for each postal code**

In [6]:
geo_df=pd.read_csv('http://cocl.us/Geospatial_data')
geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [7]:
df_toronto = pd.merge(df, geo_df, on='Postal Code')
df_toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## Part 3

Finally, we will explore and cluster using Foursquare. For this question we will narrow down the data set and look at the Toronto city area.

**Import relevant libraries**

In [8]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0          conda-forge
    geopy:           

**First we map all the neighbourhoods in the greater Toronto area**

In [9]:
address = 'Toronto, ON, Canada'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [10]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

**Next we narrow down our data set to explore the boroughs that include the name Toronto**

In [11]:
df_tor=df_toronto[df_toronto['Borough'].str.contains("Toronto")]
df_tor.head()

Postal Code           Borough                                 Neighborhood  \
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
15         M5C  Downtown Toronto                               St. James Town   
19         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  
2   43.654260 -79.360636  
4   43.662301 -79.389494  
9   43.657162 -79.378937  
15  43.651494 -79.375418  
19  43.676357 -79.293031

**We will now explore these neighbourhoods using Foursquare and examine the different types of venues prevalent in each area**

In [68]:
# The code was removed by Watson Studio for sharing.

In [13]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    radius=500
    LIMIT=100
    venues_list=[]
    
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [14]:
tor_venues = getNearbyVenues(names=df_tor['Neighborhood'],
                                   latitudes=df_tor['Latitude'],
                                   longitudes=df_tor['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [15]:
print(tor_venues.shape)
tor_venues.head()

(1622, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                      Venue  Venue Latitude  Venue Longitude  \
0          Roselle Desserts       43.653447       -79.362017   
1             Tandem Coffee       43.653559       -79.361809   
2    Cooper Koo Family YMCA       43.653249       -79.358008   
3       Body Blitz Spa East       43.654735       -79.359874   
4  Dominion Pub and Kitchen       43.656919       -79.358967   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4                  Pub

**Here we assess how many venues are listed for each neighbourhood** \
We find that there are a total of 1,622 venues and 233 unique categories of venues.

In [35]:
tor_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton, Parkdale Village, Exhibition Place                           24   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     65   
Christie                                                               16   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           14   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         24   
Kensington Market, Chinatown, Grange Park                              59   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             2   
North Toronto West, Lawrence Park                                      19   
Parkdale, Roncesvalles                                                 15   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     37   
St. James Town                                                         80   
St. James Town, Cabbagetown                                            44   
Stn A PO Boxes                                                         97   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton, Parkdale Village, Exhibition Place                            24   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      65   
Christie                                                                16   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              35   
Davisville North                    

In [36]:
print('There are {} uniques categories.'.format(len(tor_venues['Venue Category'].unique())))

There are 233 uniques categories.


In [38]:
# one hot encoding
tor_onehot = pd.get_dummies(tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
tor_onehot['Neighborhood'] = tor_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]

tor_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0                 0              0          0           0           0       1   
1                 0              0          0           0           0       0   
2                 0              0          0           0           0       0   
3                 0              0          0           0           0       0   
4                 0              0          0           0           0       0   

   Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  \
0     0    0                 0                   0      0                0   
1     0    0                 0                   0      0                0   
2     0    0                 0                   0      0                0   
3     0    0                 0                   0      0                0   
4     0    0                 0                   0      0                0   

   Beer Bar  Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  \
0         0           0                   0       0              0          0   
1         0           0                   0       0              0          0   
2         0           0                   0       0              0          0   
3         0           0                   0       0              0          0   
4         0           0                   0       0              0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0                     0               0        0                0   
1         0                     0               0        0                0   
2         0                     0               0        0                0   
3         0                     0               0        0                0   
4         0                     0               0        0                0   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0         0             0              0         0        0     0   
1         0             0              0         0        0     0   
2         0             0              0         0        0     0   
3         0             0              0         0        0     0   
4         0             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Candy Store  Caribbean Restaurant  \
0                          0             0            0                     0   
1                          0             0            0                     0   
2                          0    

In [19]:
tor_onehot.shape

(1622, 233)

**Now we assess the frequency with which each category occurs by neighbourhood**

In [20]:
tor_grouped = tor_onehot.groupby('Neighborhood').mean().reset_index()
tor_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.058824   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.015385   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.025974   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.022727   
20                        Moore Park, Summerhill East     0.000000   
21                  North Toronto West, Lawrence Park     0.052632   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.030303   
24                          Regent Park, Harbourfront     0.022222   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.027027   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010309   
32                                    Studio District     0.025000   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023256   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.028571   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.058824            0.058824      0.058824   
4            0.000000  0.000000            0.000000      0.000000   
5            0.000000  0.000000            0.000000      0.000000   
6            0.012987  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.000000  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.000000  0.000000            0.000000      0.000000  

**We will now cluster the neighbourhoods based on the most commonly occuring category of venue**

In [53]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [54]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = tor_grouped['Neighborhood']

for ind in np.arange(tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                  Café           Cheese Shop   
1                Bakery        Breakfast Spot           Coffee Shop   
2         Auto Workshop            Skate Park    Light Rail Station   
3        Airport Lounge      Airport Terminal      Sculpture Garden   
4    Italian Restaurant        Sandwich Place   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant            Restaurant   
1         Grocery Store               Stadium         Burrito Place   
2         Burrito Place        Farmers Market  Fast Food Restaurant   
3              Boutique   Rental Car Location                 Plane   
4                  Café                   Bar          Burger Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar                Lounge                   Park  
1            Restaurant          Climbing Gym              Pet Store  
2               Butcher            Restaurant       Recording Studio  
3           Coffee Shop         Boat or Ferry        Harbor / Marina  
4           Salad Place       Thai Restaurant       Department Store

In [55]:
# set number of clusters
kclusters = 3

tor_grouped_clustering = tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [56]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [57]:
tor_merged = df_tor

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
tor_merged = tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

tor_merged.dropna(subset=["Cluster Labels"], axis=0, inplace=True)
tor_merged['Cluster Labels'] = tor_merged['Cluster Labels'].astype(int)

tor_merged.head()

Postal Code           Borough                                 Neighborhood  \
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
15         M5C  Downtown Toronto                               St. James Town   
19         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2   43.654260 -79.360636               1           Coffee Shop   
4   43.662301 -79.389494               1           Coffee Shop   
9   43.657162 -79.378937               1        Clothing Store   
15  43.651494 -79.375418               1           Coffee Shop   
19  43.676357 -79.293031               0     Health Food Store   

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
2                 Bakery                   Pub                       Park   
4       Sushi Restaurant                 Diner        Arts & Crafts Store   
9            Coffee Shop        Cosmetics Shop  Middle Eastern Restaurant   
15                  Café   American Restaurant                  Gastropub   
19                 Trail                   Pub              Women's Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
2         Breakfast Spot                  Café                      Theater   
4          Burrito Place   Distribution Center               Discount Store   
9                   Café       Bubble Tea Shop          Japanese Restaurant   
15          Cocktail Bar            Restaurant                     Creperie   
19          Dance Studio     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2            Yoga Studio        Cosmetics Shop             Shoe Store  
4          Smoothie Shop                  Café               Beer Bar  
9            Pizza Place                Bakery              Bookstore  
15        Cosmetics Shop        Clothing Store         Lingerie Store  
19            Donut Shop      Doner Restaurant                Dog Run

In [58]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighborhood'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters


### Cluster 1

This is the largest cluster with the most common venue category being coffee shops.

In [67]:
tor_merged.loc[tor_merged['Cluster Labels'] == 1, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Downtown Toronto               1           Coffee Shop   
4    Downtown Toronto               1           Coffee Shop   
9    Downtown Toronto               1        Clothing Store   
15   Downtown Toronto               1           Coffee Shop   
20   Downtown Toronto               1           Coffee Shop   
24   Downtown Toronto               1           Coffee Shop   
25   Downtown Toronto               1         Grocery Store   
30   Downtown Toronto               1           Coffee Shop   
31       West Toronto               1              Pharmacy   
36   Downtown Toronto               1           Coffee Shop   
37       West Toronto               1                   Bar   
41       East Toronto               1      Greek Restaurant   
42   Downtown Toronto               1           Coffee Shop   
43       West Toronto               1                  Café   
47       East Toronto               1                  Park   
48   Downtown Toronto               1           Coffee Shop   
54       East Toronto               1                  Café   
61    Central Toronto               1                  Park   
67    Central Toronto               1                  Park   
69       West Toronto               1    Mexican Restaurant   
73    Central Toronto               1        Clothing Store   
74    Central Toronto               1           Coffee Shop   
75       West Toronto               1             Gift Shop   
79    Central Toronto               1           Pizza Place   
80   Downtown Toronto               1                  Café   
81       West Toronto               1           Coffee Shop   
84   Downtown Toronto               1                  Café   
86    Central Toronto               1                   Pub   
87   Downtown Toronto               1       Airport Service   
92   Downtown Toronto               1           Coffee Shop   
96   Downtown Toronto               1           Coffee Shop   
97   Downtown Toronto               1           Coffee Shop   
99   Downtown Toronto               1           Coffee Shop   
100      East Toronto               1           Yoga Studio   

    2nd Most Common Venue  3rd Most Common Venue  \
2                  Bakery                    Pub   
4        Sushi Restaurant                  Diner   
9             Coffee Shop         Cosmetics Shop   
15                   Café    American Restaurant   
20           Cocktail Bar                   Café   
24     Italian Restaurant         Sandwich Place   
25                   Café                   Park   
30             Restaurant                   Café   
31                 Bakery          Grocery Store   
36               Aquarium                  Hotel   
37             Restaurant  Vietnamese Restaurant   
41            Coffee Shop     Italian Restaurant   
42                  Hotel                   Café   
43                 Bakery         Breakfast Spot   
47         Sandwich Place   Fast Food Restaurant   
48                   Café             Restaurant   
54            Coffee Shop              Gastropub   
61            Swim School               Bus Line   
67                    Gym         Breakfast Spot   
69                   Café        Thai Restaurant   
73            Coffee Shop          Grocery Store   
74                   Café         Sandwich Place   
75         Breakfast Spot              Bookstore   
79           Dessert Shop         Sandwich Place   
80                 Bakery              Bookstore   
81       Sushi Restaurant                   Café   
84     Mexican Restaurant  Vietnamese Restaurant   
86            Coffee Shop    Fried Chicken Joint   
87         Airport Lounge       Airport Terminal   
92     Seafood Restaurant                   Café   
96            Pizza Place             Restaurant   
97                   Café             Restaurant   
99       Sushi Restaurant    Japanese Restaurant   
100         Auto Workshop             Sk

### Cluster 2
The second cluster is significantly smaller with Trails being the most common venue category

In [65]:
tor_merged.loc[tor_merged['Cluster Labels'] == 0, tor_merged.columns[[0] + list(range(5, tor_merged.shape[1]))]]

Postal Code  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19         M4E               0     Health Food Store                 Trail   
68         M5P               0                  Park         Jewelry Store   
83         M4T               0                   Gym                 Trail   
91         M4W               0                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19                   Pub         Women's Store          Dance Studio   
68                 Trail      Sushi Restaurant          Dance Studio   
83         Women's Store          Dance Studio  Ethiopian Restaurant   
91                 Trail          Cupcake Shop     Electronics Store   

          6th Most Common Venue        7th Most Common Venue  \
19            Electronics Store  Eastern European Restaurant   
68            Electronics Store  Eastern European Restaurant   
83            Electronics Store  Eastern European Restaurant   
91  Eastern European Restaurant                   Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19            Donut Shop      Doner Restaurant                Dog Run  
68            Donut Shop      Doner Restaurant                Dog Run  
83            Donut Shop      Doner Restaurant                Dog Run  
91      Doner Restaurant               Dog Run    Distribution Center

### Cluster 3
There is only one neighbourhood in the third cluster with Garden being the most common venue

In [61]:
tor_merged.loc[tor_merged['Cluster Labels'] == 2, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto               2                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62           Music Venue         Women's Store          Dance Studio   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
62     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62      Doner Restaurant               Dog Run    Distribution Center